### PR TITLE
MOD-8144: Restore filtering stop words in the beginning of tag queries

### PR DESCRIPTION
**Pull request by Itzikvaknin**
Fix stopword filtering for the first two terms in tag queries.

Previous behavior:

```
127.0.0.1:6379> FT.EXPLAINCLI idx '@t2:{with as the an}' dialect 2
TAG:@t2 {
  INTERSECT {
    with
    as
  }
}
```

New behavior:

```
127.0.0.1:6379> FT.EXPLAINCLI idx '@t2:{with as the an}' dialect 2
<empty>
```

This change is consistent with the text query:

```
127.0.0.1:6379> FT.EXPLAINCLI idx 'as with' dialect 2
<empty>
```

related to https://github.com/RediSearch/RediSearch/pull/3130

### CODE REVIEW

File: tests/pytests/test_parser.py [hunk 1]
```diff
@@ -617,3 +617,13 @@ def testModifierList(env):
   }
 }
 '''[1:])
+
+def testTagQueryWithStopwords_V2(env):
+    env = Env(moduleArgs = 'DEFAULT_DIALECT 2')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 't1', 'TEXT', 't2', 'TAG').ok()
+    env.expect('FT.EXPLAIN', 'idx', '@t2:{as is the with by}').equal(r'''
```
Comment (oshadmi) [inline tests/pytests/test_parser.py:624]: Add test with none-stopword first or second terms to see they are not filtered

**Review by nafraf — APPROVED** [review]

**Review by oshadmi — DISMISSED** [review]

File: tests/pytests/test_parser.py [hunk 2]
```diff
@@ -629,6 +629,28 @@ def testModifierList(env):
 }
 '''[1:])
 
+def testQuotes_v2():
```
Comment (oshadmi) [inline tests/pytests/test_parser.py:632]: Is this related to pull #5568 ? Not part of this PR?

Comment (Itzikvaknin) [reply tests/pytests/test_parser.py:632]: Yes

File: tests/pytests/test_parser.py [hunk 3]
```diff
@@ -650,3 +650,26 @@ def testQuotes_v2():
         env.expect('FT.EXPLAIN', 'idx', f'\'{query}\'').equal(expected)
         squote_query = query.replace('"', '\'')
         env.expect('FT.EXPLAIN', 'idx', f'"{squote_query}"').equal(expected)
+
+def testTagQueryWithStopwords_V2(env):
+    env = Env(moduleArgs = 'DEFAULT_DIALECT 2')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 't1', 'TEXT', 't2', 'TAG').ok()
```
Comment (raz-mon) [inline tests/pytests/test_parser.py:656]: Remove the `TEXT` field if it's not used

File: tests/pytests/test_parser.py [hunk 4]
```diff
@@ -650,3 +650,26 @@ def testQuotes_v2():
...
+    env.expect('FT.EXPLAIN', 'idx', '@t2:{as is the with by}').equal(r'''
```
[the remark this answers is not in the document]
Comment (Itzikvaknin) [reply tests/pytests/test_parser.py:657]: To have a reasonable test for FT.SEARCH, the misalignment between FT.EXPLAINCLI and FT.SEARCH needs to be fixed. The test is written and commented out until then.

File: src/query_parser/v2/parser.y [hunk 5]
```diff
@@ -600,8 +600,12 @@ text_expr(A) ::= verbatim(B) . [VERBATIM]  {
...
 termlist(A) ::= param_term(B) param_term(C). [TERMLIST]  {
   A = NewPhraseNode(0);
-  QueryNode_AddChild(A, NewTokenNode_WithParams(ctx, &B));
-  QueryNode_AddChild(A, NewTokenNode_WithParams(ctx, &C));
+  if (!(B.type == QT_TERM && StopWordList_Contains(ctx->opts->stopwords, B.s, B.len))) {
+    QueryNode_AddChild(A, NewTokenNode_WithParams(ctx, &B));
+  }
+  if (!(C.type == QT_TERM && StopWordList_Contains(ctx->opts->stopwords, C.s, C.len))) {
+    QueryNode_AddChild(A, NewTokenNode_WithParams(ctx, &C));
+  }
```
[the remark this answers is not in the document]
Comment (Itzikvaknin) [reply src/query_parser/v2/parser.y]: This rule should be looked at more broadly since it creates a phrase node that can combine more than two terms when deriving this rule: 
termlist(A) ::= termlist(B) param_term(C).

**Review by oshadmi — DISMISSED** [review]

**Comment by Itzikvaknin** [comment]
related to #3130

File: tests/pytests/test_parser.py [hunk 6]
```diff
@@ -650,3 +650,26 @@ def testQuotes_v2():
...
+    cat
+    dog
+  }
+}
+'''[1:])
+    conn = env.getClusterConnectionIfNeeded()
+    conn.execute_command('HSET', 'doc1', 't', 'with')
+    conn.execute_command('HSET', 'doc2', 't', 'cat dog')
+    env.expect('FT.SEARCH', 'idx', '@t:{cat with dog}').equal([1, 'doc2', ['t', 'cat dog']])
+    env.expect('FT.SEARCH', 'idx', '@t:{as is the with by}').equal([0])
```
Comment (raz-mon) [inline tests/pytests/test_parser.py:675]: End with new line

File: src/query_parser/v2/parser.y (hunk 5 above)
[the remark this answers is not in the document]
Comment (raz-mon) [reply src/query_parser/v2/parser.y:608]: Not handling this case may introduce some unexpected issues later on, let's handle it explicitly (here or in the AST evaluator, preferably here)

File: src/query_parser/v2/parser.y [hunk 7]
```diff
@@ -800,28 +804,55 @@ tag_list(A) ::= verbatim(B) . [TAGLIST] {
...
 
 tag_list(A) ::= tag_list(B) OR affix(C) . [TAGLIST] {
-  QueryNode_AddChild(B, C);
-  A = B;
+  if (!B){
+    QueryNode_AddChild(B, C);
+    A = B;
+  } else {
+    A = NewPhraseNode(0);
+    QueryNode_AddChild(A, C);
```
Comment (raz-mon) [inline src/query_parser/v2/parser.y:832]: Where's `B`?

File: src/query_parser/v2/parser.y [hunk 8]
```diff
@@ -800,28 +804,55 @@ tag_list(A) ::= verbatim(B) . [TAGLIST] {
...
 
 tag_list(A) ::= tag_list(B) OR verbatim(C) . [TAGLIST] {
-  QueryNode_AddChild(B, C);
-  A = B;
+  if (!B){
+    QueryNode_AddChild(B, C);
+    A = B;
+  } else {
+    A = NewPhraseNode(0);
+    QueryNode_AddChild(A, C);
```
Comment (raz-mon) [inline src/query_parser/v2/parser.y:842]: Also no `B` - let's add unit-tests that would have failed for this please

File: src/query_parser/v2/parser.y [hunk 9]
```diff
@@ -800,28 +804,55 @@ tag_list(A) ::= verbatim(B) . [TAGLIST] {
...
+  }
 }
 
 tag_list(A) ::= tag_list(B) OR termlist(C) . [TAGLIST] {
-  QueryNode_AddChild(B, C);
-  A = B;
+  int rv = one_not_null(B, C, (void**)&A);
```
Comment (raz-mon) [inline src/query_parser/v2/parser.y:847]: `termlist` cannot be `NULL`, no need to handle that case.

File: src/query_parser/v2/parser.y [hunk 10]
```diff
@@ -800,28 +804,53 @@ tag_list(A) ::= verbatim(B) . [TAGLIST] {
 }
 
 tag_list(A) ::= termlist(B) . [TAGLIST] {
-  A = NewPhraseNode(0);
-  QueryNode_AddChild(A, B);
+  if (QueryNode_NumChildren(B) == 0 ){
```
Comment (oshadmi) [inline src/query_parser/v2/parser.y:807]: Can use likely or __builtin_expect since this branch would rarely be taken

File: src/query_parser/v2/parser.y [hunk 11]
```diff
@@ -800,28 +804,53 @@ tag_list(A) ::= verbatim(B) . [TAGLIST] {
...
+  } else {
+    A = NewPhraseNode(0);
+    QueryNode_AddChild(A, B);
+  }
 }
 
 tag_list(A) ::= tag_list(B) OR param_term_case(C) . [TAGLIST] {
-  QueryNode_AddChild(B, NewTokenNode_WithParams(ctx, &C));
-  A = B;
+  if (!B){
```
Comment (oshadmi) [inline src/query_parser/v2/parser.y:817]: Consider also NULL `tag_list` is very unlikely and would usually not be taken

File: tests/pytests/test_parser.py [hunk 12]
```diff
@@ -650,3 +650,76 @@ def testQuotes_v2():
...
+ # tag_list ::= taglist OR affix (affix is suffix)
+  env.expect('FT.EXPLAIN', 'idx', '@tag:{x y | *a }').equal(r'''
+TAG:@tag {
+  INTERSECT {
+    x
+    y
+  }
+  SUFFIX{*a}
+}
+'''[1:])
```
Comment (raz-mon) [inline tests/pytests/test_parser.py:692]: Let's use a prefix matching the search query below (`*ple`)

File: tests/pytests/test_parser.py [hunk 13]
```diff
@@ -650,3 +650,76 @@ def testQuotes_v2():
...
+  env.expect('FT.SEARCH', 'idx', '@tag:{as with | *na }').equal([1, 'doc3', ['tag', 'banana']])
+
+  # tag_list ::= taglist OR affix (affix is prefix)
+  env.expect('FT.EXPLAIN', 'idx', '@tag:{x y | b* }').equal(r'''
+TAG:@tag {
+  INTERSECT {
+    x
+    y
+  }
+  PREFIX{b*}
```
Comment (raz-mon) [inline tests/pytests/test_parser.py:704]: Weird that we don't have `UNION` here.. Only an explain bug, or in the parser?

File: tests/pytests/test_parser.py [hunk 14]
```diff
@@ -650,3 +650,76 @@ def testQuotes_v2():
...
+  }
+  INFIX{*n*}
+}
+'''[1:])
+  env.expect('FT.SEARCH', 'idx', '@tag:{x y | *pl* }').equal([2, 'doc1', ['tag', 'x y'], 'doc2', ['tag', 'apple']])
+
+# taglist OR param_term_case
+  env.expect('FT.SEARCH', 'idx', '@tag:{x y | banana }').equal([2, 'doc1', ['tag', 'x y'], 'doc3', ['tag', 'banana']])
+# test tag_list ::= taglist OR param_term_case with an empty taglist on the RHS
+  env.expect('FT.SEARCH', 'idx', '@tag:{as with | banana }').equal([1, 'doc3', ['tag', 'banana']])
```
Comment (raz-mon) [inline tests/pytests/test_parser.py:724]: Let's add the explain

**Review by raz-mon — COMMENTED** [review]
Almost there

File: tests/pytests/test_parser.py (hunk 13 above)
Comment (Itzikvaknin) [reply tests/pytests/test_parser.py:704]: Only an explain bug

**Review by raz-mon — APPROVED** [review]

**Review by oshadmi — APPROVED** [review]
